pkg/sources/github/github.go

```diff
@@ -1128,13 +1132,29 @@ func getRepoURLParts(repoURL string) (string, []string, error) {
 	urlString := url.String()
 	trimmedURL := strings.TrimPrefix(urlString, url.Scheme+"://")
 	trimmedURL = strings.TrimSuffix(trimmedURL, ".git")
-	splitURL := strings.Split(trimmedURL, "/")
+	urlParts := strings.Split(trimmedURL, "/")
 
-	if len(splitURL) < 2 || len(splitURL) > 3 {
-		return "", []string{}, fmt.Errorf("invalid repository or gist URL (%s): length of URL segments should be 2 or 3", urlString)
+	// Validate
+	switch len(urlParts) {
+	case 2:
+		if !strings.EqualFold(urlParts[0], "gist.github.com") {
+			err = fmt.Errorf("failed to parse repository or gist URL (%s): 2 segments are only expected if the host is 'gist.github.com'", urlString)
+		}
+	case 3:
+		// Do nothing
+	case 4:
+		if !(!strings.EqualFold(urlParts[0], "github.com") && strings.EqualFold(urlParts[1], "gist")) {
```

Comment:
I think there's an extra `!` in there and it should be:

```go
  !(strings.EqualFold(urlParts[0], "github.com") && strings.EqualFold(urlParts[1], "gist"))
// ^^ removed this one
```

A `gist` segment is only expected for the `github.com` host, right?

Reply:
This is difficult to grok in its current state, I'll need to change it — or at least add a comment. 😵 

> I think there's an extra `!` in there and it should be:
> 
> ```go
>   !(strings.EqualFold(urlParts[0], "github.com") && strings.EqualFold(urlParts[1], "gist"))
> // ^^ removed this one
> ```
> 
> A `gist` segment is only expected for the `github.com` host, right?

It's actually the opposite. GitHub uses the subdomain `gist.github.com` for gists, and although `github.com/gist` redirects -> `gist.github.com`, it isn't a valid gist URL (e.g., https://github.com/gist/280192b0523099d0614a95f579d99aa9 and https://github.com/gist/rgmz/280192b0523099d0614a95f579d99aa9). However, `/gist/<id>` can be valid for instances of GitHub Enterprise Server (e.g., https://github.company.org/gist/nat/5fdbb7f945d121f197fb074578e53948.git).

The logic is meant to accept a `gist` segment for any domain _except_ `github.com`.

Reply:
I'm not sure if @joeleonjr has anything to add about Gist URLs.

Reply:
> This is difficult to grok in its current state, I'll need to change it — or at least add a comment. 😵

I did both. Let me know if that looks better.